### PR TITLE
IO scheduling on HTTP session level

### LINKED ISSUE #63213: Smaller granularity of resource requests for S3 buffers
S3 reads and writes issued from `WriteBufferFromS3` and `ReadBufferFromS3` are done under `ResourceGuard`. But the number of bytes written (or read) in a single `PutObject` (or `GetObject`) AWS request can be high: tens or hundreds of MBs. Huge `cost`s lead to larger unfairness and `bandwidth_limit` [throttling](https://clickhouse.com/docs/en/operations/workload-scheduling#hierarchy) violation in case if `max_burst` is comparable or less than `cost` of a single resource request.

**Describe the solution you'd like**
Probably, integration with IO Scheduler should be moved from S3 Buffers to S3 Client. Consider reimplementing `ResourceGuard` as `RateLimiterInterface` implementation.

### PR DESCRIPTION
**Pull request by serxa**
Fixes https://github.com/ClickHouse/ClickHouse/issues/63213

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
IO scheduling for remote S3 disks is now done on the level of HTTP socket streams (instead of the whole S3 requests) to resolve `bandwidth_limit` throttling issues.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: Normal Builds
- [ ]  Allow: Special Builds
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 19c2f7109ffe98b5a6b556d4be0e950f510a28d5 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65182/19c2f7109ffe98b5a6b556d4be0e950f510a28d5/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: base/poco/Net/src/SocketImpl.cpp
Comment (serxa) [inline base/poco/Net/src/SocketImpl.cpp:1]: It is just spaces because I ended up not changing this file. We can revert the changes here if necessary.

File: src/Common/CurrentThread.h [hunk 1]
```diff
@@ -102,6 +110,39 @@ class CurrentThread
         void logPeakMemoryUsage();
         bool log_peak_memory_usage_in_destructor = true;
     };
+
+    /// Scoped attach/detach of IO resource links
+    struct IOScope : private boost::noncopyable
```
Comment (serxa) [inline src/Common/CurrentThread.h:115]: I think we will use this scope eventually for all IO operations including local, hence the name.

File: src/Common/HTTPConnectionPool.cpp [hunk 2]
```diff
@@ -236,6 +237,51 @@ class IExtendedPool : public IHTTPConnectionPoolForEndpoint
...
+        request.finish(bytes, link);
+    }
+
+    void fail() override
+    {
+        request.finish(0, link);
+    }
+
+    ResourceLink link;
+    ResourceGuard::Request request;
```
Comment (serxa) [inline src/Common/HTTPConnectionPool.cpp:281]: Technically we could have used TLS request as `ResourceGuard` does and avoid any allocations, but it feels just safer and more straightforward to me.

File: src/Common/HTTPConnectionPool.cpp [hunk 3]
```diff
@@ -337,6 +381,19 @@ class EndpointConnectionPool : public std::enable_shared_from_this<EndpointConne
...
+            if (ResourceLink link = CurrentThread::getReadResourceLink()) {
+                Session::setReceiveDataHooks(std::make_shared<ResourceGuardSessionDataHooks>(ResourceGuard::Metrics::getIORead(), link));
+            } else {
+                Session::setReceiveDataHooks();
+            }
+            if (ResourceLink link = CurrentThread::getWriteResourceLink()) {
+                Session::setSendDataHooks(std::make_shared<ResourceGuardSessionDataHooks>(ResourceGuard::Metrics::getIOWrite(), link));
+            } else {
+                Session::setSendDataHooks();
+            }
```
Comment (serxa) [inline src/Common/HTTPConnectionPool.cpp:395]: I reset callbacks here because if we do that when we release the session, there might be a reader still using the stream. Is this correct or could we find a better place? Hooks are not holding any smart pointers to important objects, only a raw pointer, but when the query is destroyed it could lead to a dangling pointer. Though we should never use it, it feels not 100% safe.

File: src/Common/HTTPConnectionPool.cpp [hunk 4]
```diff
@@ -393,6 +450,9 @@ class EndpointConnectionPool : public std::enable_shared_from_this<EndpointConne
                 }
             }
             response_stream = nullptr;
+            // FIXME: We are not sure that response stream is fully read at this moment, so hooks could possible be called after this point, right?
+            // Session::setSendDataHooks();
+            // Session::setReceiveDataHooks();
```
Comment (serxa) [inline src/Common/HTTPConnectionPool.cpp:455]:
```suggestion
```

File: src/Common/Scheduler/ISchedulerQueue.h [hunk 5]
```diff
@@ -34,18 +37,6 @@ class ISchedulerQueue : public ISchedulerNode
...
-        adjustBudget(0, cost);
-    }
-
-    // Adjust budget to account for requested, but not consumed `cost` resource units
-    void accumulateBudget(ResourceCost cost)
-    {
-        adjustBudget(cost, 0);
-    }
-
     /// Enqueue new request to be executed using underlying resource.
```
Comment (serxa) [inline src/Common/Scheduler/ISchedulerQueue.h:40]: It is not needed anymore, now logic integrated into ResourceGuard

File: src/Common/Scheduler/ResourceGuard.h [hunk 6]
```diff
@@ -63,43 +113,59 @@ class ResourceGuard
...
         }
 
-        void finish()
+        void finish(ResourceCost real_cost_, ResourceLink link_)
         {
             // lock(mutex) is not required because `Dequeued` request cannot be used by the scheduler thread
             chassert(state == Dequeued);
             state = Finished;
+            if (estimated_cost != real_cost_)
+                link_.queue->adjustBudget(estimated_cost, real_cost_);
```
Comment (serxa) [inline src/Common/Scheduler/ResourceGuard.h:128]: There was a bug. All ResourceGuard were used in the wrong way passing `request->cost` into `adjustBudget()` but `enqueueRequestUsingBudget` changes `request->cost` before enqueing the request, so the correct initial estimated cost was not stored anywhere, but it is required for correct adjustment of the budget. Now we store it in ResourceGuard and it's usage is much simpler and less error-prone.

File: base/poco/Net/include/Poco/Net/HTTPSession.h [hunk 7]
```diff
@@ -33,6 +35,27 @@ namespace Net
...
 
 
+    class IHTTPSessionDataHooks
+    /// Interface to control stream of data bytes being sent or received though socket by HTTPSession
+    /// It allows to monitor, throttle and schedule data streams with syscall granulatrity
+    {
+    public:
+        virtual ~IHTTPSessionDataHooks() = default;
+
+        virtual void start(int bytes) = 0;
```
Comment (CheSema) [inline base/poco/Net/include/Poco/Net/HTTPSession.h:45]: I suggest to use those names for hooks.
`atStart`, `atFinish`, `atFail`

File: src/Common/CurrentThread.h (hunk 1 above)
Comment (CheSema) [reply src/Common/CurrentThread.h:115]: Do we haves cases with only one type of resources attached? Or they are used together always?

File: src/Disks/IO/ReadBufferFromAzureBlobStorage.cpp [hunk 8]
```diff
@@ -113,7 +114,9 @@ bool ReadBufferFromAzureBlobStorage::nextImpl()
     {
         try
         {
+            ResourceGuard rlock(ResourceGuard::Metrics::getIORead(), read_settings.io_scheduling.read_resource_link, to_read_bytes);
             bytes_read = data_stream->ReadToCount(reinterpret_cast<uint8_t *>(data_ptr), to_read_bytes);
+            rlock.unlock(bytes_read); // Do not hold resource under bandwidth throttler
```
Comment (CheSema) [inline src/Disks/IO/ReadBufferFromAzureBlobStorage.cpp:119]: I do not understand that. 
I thought that if we use guards we set them for some wide scope and do not bother with one line code.

File: src/Common/HTTPConnectionPool.cpp (hunk 4 above)
Comment (CheSema) [reply src/Common/HTTPConnectionPool.cpp:454]: No, they could not. Connection is destroyed. I do not find code where they are preserved for the next reuse, which is right.

File: src/Disks/IO/ReadBufferFromAzureBlobStorage.cpp (hunk 8 above)
Comment (CheSema) [reply src/Disks/IO/ReadBufferFromAzureBlobStorage.cpp:119]: I mixed it up with IOScope. `ResourceGuard` does scheduling itself.

File: src/Common/HTTPConnectionPool.cpp [hunk 9]
```diff
@@ -236,6 +237,51 @@ class IExtendedPool : public IHTTPConnectionPoolForEndpoint
...
+
+    ~ResourceGuardSessionDataHooks() override
+    {
+        request.assertFinished(); // Never destruct with an active request
+    }
+
+    void start(int bytes) override
+    {
+        request.enqueue(bytes, link);
+        request.wait();
```
[the remark this answers is not in the document]
Comment (serxa) [reply src/Common/HTTPConnectionPool.cpp:267]: Yes, it is known issue. With fair, limiting and throttling nodes scheduler will make perfectly paced schedule (as much as it possible with 1MB requests). Long delays (i.e starvation) will only arise if priority node is used. So I'd make a note about it in documentation. With local IO or CPU scheduling it should note be an issue, so priority node is useful in general case.

File: src/IO/WriteBufferFromS3.cpp [hunk 10]
```diff
@@ -700,12 +697,11 @@ void WriteBufferFromS3::makeSinglepartUpload(WriteBufferFromS3::PartData && data
             if (client_ptr->isClientForDisk())
                 ProfileEvents::increment(ProfileEvents::DiskS3PutObject);
 
-            ResourceCost cost = request.GetContentLength();
-            ResourceGuard rlock(write_settings.resource_link, cost);
+            CurrentThread::IOScope io_scope(write_settings.io_scheduling);
```
Comment (CheSema) [inline src/IO/WriteBufferFromS3.cpp:700]: Here you are controlling only particular current tread with the exact call. In other words you enable scheduling only for PUT (single object + multi parts) and GET (get object).
Also you manually set resources guards for Azure and HDFS.

But what about the rest? I expected that you cover here HTTP storages as well with other types of requests.

File: src/Common/HTTPConnectionPool.cpp [hunk 11]
```diff
@@ -337,6 +381,18 @@ class EndpointConnectionPool : public std::enable_shared_from_this<EndpointConne
         std::ostream & sendRequest(Poco::Net::HTTPRequest & request) override
         {
             auto idle = idleTime();
+
+            // Reset data hooks for IO scheduling
+            if (ResourceLink link = CurrentThread::getReadResourceLink())
```
[the remark this answers is not in the document]
Comment (CheSema) [reply src/Common/HTTPConnectionPool.cpp:386]: false warning. In that case hooks are preservers with session object.

File: src/Common/HTTPConnectionPool.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (serxa) [reply src/Common/HTTPConnectionPool.cpp:267]: Let's add logging in `ResourceGuardSessionDataHooks` if resource request takes more than 5 seconds to be dequeued from the scheduler

File: src/IO/WriteBufferFromS3.cpp (hunk 10 above)
Comment (serxa) [reply src/IO/WriteBufferFromS3.cpp:716]: Yes, it should cover all remote and local IO eventually. In this PR I'm just trying to improve what we already have. Let's also do the following in a separate PR:
1. move `CurrentThread::IOScope` into  ObjectStorage
2. make IOScope inheritable when spawning new threads 
3. get rid of `ReadSettings` and `WriteSettings` fields for passing `io_scheduling` field
It should make the process of integration other buffers and scheduler more straightforward

File: src/Common/CurrentThread.h (hunk 1 above)
Comment (serxa) [reply src/Common/CurrentThread.h:115]: For local IO it could be only read or write.

File: src/Common/HTTPConnectionPool.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (serxa) [reply src/Common/HTTPConnectionPool.cpp:271]: done

**Comment by serxa** [comment]
@maxknv Can we help build a check to finish? How can I find out what when wrong?
```
Pending — 13 of 21 builds are missing. 18/31 artifact groups are OK
```

**Comment by maxknv** [comment]
> @maxknv Can we help build a check to finish? How can I find out what when wrong?
> 
> ```
> Pending — 13 of 21 builds are missing. 18/31 artifact groups are OK
> ```

package_aarch64 failed with OOM. - I suggest restarting it manually

**Comment by serxa** [comment]
Looks like we have data race here now https://pastila.nl/?000777c6/1c25eebbf78952d5ed8d7c42cbcca009#or9ialwCHfjZi17MNTG5Hg==

File: src/Common/HTTPConnectionPool.cpp [hunk 12]
```diff
@@ -393,6 +453,8 @@ class EndpointConnectionPool : public std::enable_shared_from_this<EndpointConne
                 }
             }
             response_stream = nullptr;
+            Session::setSendDataHooks();
+            Session::setReceiveDataHooks();
```
Comment (serxa) [inline src/Common/HTTPConnectionPool.cpp:457]: This reset lead to a data race: https://pastila.nl/?000777c6/1c25eebbf78952d5ed8d7c42cbcca009#or9ialwCHfjZi17MNTG5Hg==

So we cannot reset it here, because session could be still in use for some reason. I think I should return to the logic I used a few revisions earlier: reset hooks when the connection is reassigned to another request in `sendRequest()`.

**Comment by alexey-milovidov** [comment]
> Stateless tests (release) — Some queries hung, fail: 1, passed: 6833, skipped: 6 

https://github.com/ClickHouse/ClickHouse/issues/37686

```
clickhouse-cloud :) SELECT pull_request_number, commit_sha, check_name, event_time, message FROM merge('^text_log') WHERE message_format_string = 'Received EINTR while trying to drain a TimerDescriptor, fd {}: {}' AND event_date >= yesterday()

SELECT
    pull_request_number,
    commit_sha,
    check_name,
    event_time,
    message
FROM merge('^text_log')
WHERE (message_format_string = 'Received EINTR while trying to drain a TimerDescriptor, fd {}: {}') AND (event_date >= yesterday())

Query id: 045b3f12-86b9-437c-9ce2-deb6a972c129

   ┌─pull_request_number─┬─commit_sha───────────────────────────────┬─check_name────────────────┬──────────event_time─┬─message─────────────────────────────────────────────────────────────────────────────┐
1. │               65182 │ 3bff7ddcf8891d091bc5be2b827172029fb8b76f │ Stateless tests (release) │ 2024-08-02 18:17:39 │ Received EINTR while trying to drain a TimerDescriptor, fd 84: anon_inode:[timerfd] │
   └─────────────────────┴──────────────────────────────────────────┴───────────────────────────┴─────────────────────┴─────────────────────────────────────────────────────────────────────────────────────┘
   ┌─pull_request_number─┬─commit_sha───────────────────────────────┬─check_name────────────────┬──────────event_time─┬─message─────────────────────────────────────────────────────────────────────────────┐
2. │               65182 │ 3bff7ddcf8891d091bc5be2b827172029fb8b76f │ Stateless tests (release) │ 2024-08-02 18:30:47 │ Received EINTR while trying to drain a TimerDescriptor, fd 84: anon_inode:[timerfd] │
   └─────────────────────┴──────────────────────────────────────────┴───────────────────────────┴─────────────────────┴─────────────────────────────────────────────────────────────────────────────────────┘
```

**Comment by serxa** [comment]
data race is not fixed. it just moved few lines below to a different location https://pastila.nl/?000d4920/623b6fe76e20a416a7c021a625ec3d4b#BDtMc1k+7LXEcMezAHBJhA==

File: src/Common/HTTPConnectionPool.cpp (hunk 12 above)
Comment (serxa) [reply src/Common/HTTPConnectionPool.cpp:457]: It turned out to be correct. Data race was inside ResourceGuard

**Comment by serxa** [comment]
Now data race should be fixed. It was in `ResourceGuard::Request`. It turned out that the common optimization advise "signal your condintion_variable outside critical section" is wrong. The other thread that should wait on cv could finds your updated state and do not wait at all and if it then destructs the cv, then the first thread (signalling) will wake up and signal the destructed cv. Make sure you don't fall into this trap while optimizing.
```cpp
        void execute() override
        {
            {
                std::unique_lock lock(mutex);
                chassert(state == Enqueued);
                state = Dequeued;
            }
            dequeued_cv.notify_one();  // This optimization is TOTALLY WRONG. it must be under lock
        }
```

**Review by CheSema — APPROVED** [review]
